test/unit/org/apache/cassandra/net/SimulatedMessageDelivery.java

```diff
@@ -0,0 +1,333 @@
+                                    DropListener onDropped,
+                                    Scheduler scheduler,
+                                    Consumer<Throwable> onError)
+    {
+        this.self = self;
+        this.actions = actions;
+        this.reciever = reciever;
+        this.onDropped = onDropped;
+        this.scheduler = scheduler;
+        this.onError = onError;
+    }
+
+    public void stop()
+    {
+        callbacks.clear();
+        status = Status.Down;
+    }
+
+    @Override
+    public <REQ> void send(Message<REQ> message, InetAddressAndPort to)
+    {
+        message = message.withFrom(self);
+        maybeEnqueue(message, to, null);
+    }
+
+    @Override
+    public <REQ, RSP> void sendWithCallback(Message<REQ> message, InetAddressAndPort to, RequestCallback<RSP> cb)
+    {
+        message = message.withFrom(self);
+        maybeEnqueue(message, to, cb);
+    }
+
+    @Override
+    public <REQ, RSP> void sendWithCallback(Message<REQ> message, InetAddressAndPort to, RequestCallback<RSP> cb, ConnectionType specifyConnection)
+    {
+        message = message.withFrom(self);
+        maybeEnqueue(message, to, cb);
+    }
+
+    @Override
+    public <REQ, RSP> Future<Message<RSP>> sendWithResult(Message<REQ> message, InetAddressAndPort to)
+    {
+        AsyncPromise<Message<RSP>> promise = new AsyncPromise<>();
+        sendWithCallback(message, to, new RequestCallback<RSP>()
+        {
+            @Override
+            public void onResponse(Message<RSP> msg)
+            {
+                promise.trySuccess(msg);
+            }
+
+            @Override
+            public void onFailure(InetAddressAndPort from, RequestFailureReason failure)
+            {
+                promise.tryFailure(new MessagingService.FailureResponseException(from, failure));
+            }
+
+            @Override
+            public boolean invokeOnFailure()
+            {
+                return true;
+            }
+        });
+        return promise;
+    }
+
+    @Override
+    public <V> void respond(V response, Message<?> message)
+    {
+        send(message.responseWith(response), message.respondTo());
+    }
+
+    private <REQ, RSP> void maybeEnqueue(Message<REQ> message, InetAddressAndPort to, @Nullable RequestCallback<RSP> callback)
+    {
+        if (status != Status.Up)
+            return;
+        CallbackContext cb;
+        if (callback != null)
+        {
+            CallbackKey key = new CallbackKey(message.id(), to);
+            if (callbacks.containsKey(key))
+                throw new AssertionError("Message id " + message.id() + " to " + to + " already has a callback");
+            cb = new CallbackContext(callback);
+            callbacks.put(key, cb);
+        }
+        else
+        {
+            cb = null;
+        }
+        Action action = actions.get(self, message, to);
+        switch (action)
+        {
+            case DELIVER:
```

Comment:
Should we add a delay here?

Reply:
I tried to decouple so the user could define.  In repair it enqueues with a random delay, and in epoch sync it does similar... 

looking at the `MessageDeliveryTest` it just does

`ScheduledExecutorPlus scheduler`

```
scheduler::schedule
```

that is a simulated executor but the scheduling jitter is "tuned" for CPU scheduling and not network scheduling...

TL;DR - by trying to decouple I made it harder and less clear that you had to worry about this... so yeah... we should add a delay here...

Reply:
pushed.  There is now a reusable impl of

`noDelay` - for cases that don't want network delays
`randomDelay` - this creates a random delay for each direction (from/to pair; to/from has a different distribution)

When writing this I saw that this original logic was written before `Gens.mixedDistribution` functions got added... this would be a great case for `Gens.ints().mixedDistribution(MICROSECONDS.toNanos(500), SECONDS.toNano(5), buckets=100)`...   right now there are 2 buckets (small/large)... this is in-sync with repair and burn test (both use the same logic)...

Reply:
made the change.. we no longer allocate a promise unless you don't provide a `OnResult` callback... this is only used by tests atm...

Reply:
Nice! Thank you for addressing ths.